fuel-vm/src/interpreter/memory.rs

```diff
@@ -326,23 +326,74 @@ impl MemoryInstance {
         Ok(())
     }
 
-    /// Copies the memory from `src` to `dst`.
+    /// Copies the memory from `src` to `dst` verifying ownership.
     #[inline]
     #[track_caller]
-    pub fn memcopy_noownerchecks<A: ToAddr, B: ToAddr, C: ToAddr>(
+    pub fn memcopy(
         &mut self,
-        dst: A,
-        src: B,
-        len: C,
+        dst: Word,
+        src: Word,
+        length: Word,
+        owner: OwnershipRegisters,
     ) -> Result<(), PanicReason> {
-        // TODO: Optimize
+        let dst_range = self.verify(dst, length)?;
+        let src_range = self.verify(src, length)?;
+
+        if dst_range.start() <= src_range.start() && src_range.start() < dst_range.end()
+            || src_range.start() <= dst_range.start()
+                && dst_range.start() < src_range.end()
+            || dst_range.start() < src_range.end() && src_range.end() <= dst_range.end()
+            || src_range.start() < dst_range.end() && dst_range.end() <= src_range.end()
+        {
+            return Err(PanicReason::MemoryWriteOverlap)
+        }
 
-        let src = src.to_addr()?;
-        let dst = dst.to_addr()?;
-        let len = len.to_addr()?;
+        owner.verify_ownership(&dst_range)?;
+
+        if src_range.end() <= self.stack.len() {
```

Comment:
Is all the following validation logic new?

Reply:
I moved the logic from `self.read(src, len)` and `self.write_noownerchecks(dst, len)?` here to avoid `to_vec` for `tmp` slice